### PR TITLE
feat: Add Reward Model Cluster for LLM-as-Judge in RLVR Pipeline  

### PR DESCRIPTION
**Pull request by tanzelin430**
### Motivation

The existing RLVR pipeline only supports rule-based reward workers (e.g., math_verify) or external API calls. To enable **LLM-as-Judge PRM (Process Reward Model)** within the pipeline, we need a shared vLLM inference cluster that multiple CPU reward workers can call concurrently to judge solution quality.

### Architecture
image

A new **Reward Model Cluster** is added to the RLVR Pipeline, reusing the existing `Cluster` / `RequestScheduler` / `InferWorker` components (same pattern as the agentic pipeline):

- **Reward Model Cluster** (dedicated GPUs): vLLM InferWorker stays loaded permanently, independent of the train/infer offload cycle
- **LLMClusterRewardWorker** (CPU-only): decodes actor outputs → formats judge prompts → sends concurrent requests via RequestScheduler → parses scores
- **RequestScheduler** (Ray Named Actor): discoverable across nodes, sticky routing with least-loaded dispatch

### Changes

| File | Change |
|------|--------|
| `rewards/llm_cluster_reward_worker.py` | **NEW** — CPU-only reward worker |
| `examples/test-reward-model-cluster/` | **NEW** — 8-GPU test config |
| `rlvr_config.py` | +6 lines — add `reward_model` field |
| `rlvr_pipeline.py` | +36 lines — create Cluster + RequestScheduler + init |
| `rewards/__init__.py` | +1 line — import |

### Compatibility

- **Zero intrusion**: all new code paths are guarded by `if self.reward_model_cluster:`; existing pipelines are unaffected when `reward_model` is not configured
- **Sync / Async compatible**: reward model GPUs never overlap with train/infer GPUs
- **Multi-node safe**: scheduler discovered via `ray.get_actor(name, namespace)` across the Ray cluster

### Test

Passed on 8-GPU single node (Qwen3-4B as both actor and judge). Full loop verified: generate → reward via cluster → GRPO train step. Test config: [`examples/test-reward-model-cluster/rlvr_config_test.yaml`](../examples/test-reward-model-cluster/rlvr_config_test.yaml).

### CODE REVIEW

File: examples/test-reward-model-cluster/rlvr_config_test.yaml [hunk 1]
```diff
@@ -0,0 +1,160 @@
+defaults:
```
Comment (PanAndy) [inline examples/test-reward-model-cluster/rlvr_config_test.yaml]: Please move the sample configuration file to `examples/qwen2.5-7B-rlvr_megatron/rlvr_config_8gpus_llm_as_judge_server.yaml`.

Using the data in the ROLL repository, I tested the following YAML file:


```yaml
hydra:
  run:
    dir: .
  output_subdir: null

exp_name: "qwen2.5-7B-rlvr-config"
seed: 42
logging_dir: ./output/logs
output_dir: ./output
system_envs:
  USE_MODELSCOPE: '1'

checkpoint_config:
  type: file_system
  output_dir: /data/cpfs_0/rl_examples/models/${exp_name}

#track_with: wandb
#tracker_kwargs:
#  api_key:
#  project: roll_examples
#  notes: roll_examples
#  tags:
#    - rlvr
#    - baseline

track_with: tensorboard
tracker_kwargs:
  log_dir: /data/oss_bucket_0/rl_examples/llm/tensorboard/roll_exp/rlvr

num_gpus_per_node: 8

max_steps: 500
save_steps: 100
logging_steps: 1
eval_steps: 10
resume_from_checkpoint: false
async_generation_ratio: 0

rollout_batch_size: 64  # prompt
prompt_length: 2048
response_length: 4096

num_return_sequences_in_group: 8
ppo_epochs: 1
adv_estimator: "reinforce"

# clip
value_clip: 0.5
reward_clip: 10
advantage_clip: 2.0
dual_clip_loss: true

# normalize
norm_mean_type: ~
norm_std_type: ~

# data mask
max_len_mask: true
difficulty_mask: true
difficulty_low_threshold: 0.1
difficulty_high_threshold: 0.95
error_max_len_clip: false

# data weight
difficulty_loss_weight: false
length_loss_weight: false

# reward
add_token_level_kl: false

# advantage
whiten_advantages: true

# dynamic sampling scheduler
# use_additional_prompts: true
# max_running_requests: 256
# is_num_return_sequences_expand: false
global_template: qwen2_5

pretrain: Qwen/Qwen2.5-7B
reward_pretrain: Qwen/Qwen2.5-7B

# validation:
#   data_args:
#     template: qwen2_5
#     file_name:
#       - data/math_benchmarks.jsonl
#   generating_args:
#     max_new_tokens: ${response_length}
#     top_p: 0.6
#     top_k: 50
#     num_beams: 1
#     temperature: 0.6
#     num_return_sequences: 1


actor_train:
  model_args:
    disable_gradient_checkpointing: false
    dtype: bf16
    model_type: ~
  training_args:
    learning_rate: 1.0e-6
    weight_decay: 0
    per_device_train_batch_size: 1
    gradient_accumulation_steps: 64
    warmup_steps: 20
    num_train_epochs: 50
  data_args:
    template: qwen2_5
    file_name:
      - data/llm_judge_Multi-subject-RLVR_deal_new.jsonl
    domain_interleave_probs:
      llm_judge: 1.0
    dataset_dir: data
    messages: messages
    interleave_probs: "1.0"
    preprocessing_num_workers: 16
  strategy_args:
    strategy_name: megatron_train
    strategy_config:
      tensor_model_parallel_size: 1
      pipeline_model_parallel_size: 1
      expert_model_parallel_size: 1
      use_distributed_optimizer: true
      recompute_granularity: full
  device_mapping: list(range(0,8))
  infer_batch_size: 4

actor_infer:
  model_args:
    disable_gradient_checkpointing: true
    dtype: bf16
  generating_args:
    max_new_tokens: ${response_length}
    top_p: 0.99
    top_k: 100
    num_beams: 1
    temperature: 0.99
    num_return_sequences: ${num_return_sequences_in_group}
  data_args:
    template: qwen2_5
  strategy_args:
    strategy_name: vllm
    strategy_config:
      gpu_memory_utilization: 0.8
      block_size: 16
      max_model_len: 8000
      load_format: auto
  device_mapping: list(range(0,6))
  infer_batch_size: 1

reference:
  model_args:
    disable_gradient_checkpointing: true
    dtype: bf16
    model_type: ~
  data_args:
    template: qwen2_5
  strategy_args:
    strategy_name: megatron_infer
    strategy_config:
      tensor_model_parallel_size: 1
      pipeline_model_parallel_size: 1
      expert_model_parallel_size: 1
  device_mapping: list(range(0,8))
  infer_batch_size: 8

rewards:
  llm_judge:
    worker_cls: roll.pipeline.rlvr.rewards.llm_cluster_reward_worker.LLMClusterRewardWorker
    tag_included: [RLVR]
    world_size: 32  # reward model下，可以把reward worker数量拉大，提高服务能力，GPU在reward model上，reward worker就是client了
    judge_prompt: virtuoussy/Qwen2.5-7B-Instruct-RLVR
    model_args:
      model_name_or_path: ${pretrain}
    data_args:
      template: native
    
reward_model:
  name: reward_llm
  worker_cls: roll.pipeline.base_worker.InferWorker
  model_args:
    model_name_or_path: ${rewards.llm_judge.model_args.model_name_or_path}
    attn_implementation: fa2
    dtype: bf16
  generating_args:
    max_new_tokens: 512
    temperature: 0.3
    num_return_sequences: 1
  data_args:
    template: native
  strategy_args:
    strategy_name: vllm
    strategy_config:
      gpu_memory_utilization: 0.85
      enforce_eager: false
      max_model_len: 8192
      tensor_parallel_size: 1
      load_format: auto
  device_mapping: "[6, 7]"
```

**Comment by PanAndy** [comment]
No issues now—could you rebase onto the `main` branch? 
Thanks!
